### PR TITLE
TLS certificate type tests

### PR DESCRIPTION
**Pull request by snhenson**
##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated
- [x] CLA is signed

##### Description of change

This adds options to test the certificate type returned by a server or client.

It also adds certificate selection tests: two check the certificate is selected properly when we restrict the ciphersuite and the others when we restrict the supported signature algorithms.

### CODE REVIEW

File: test/ssl-tests/20-cert-select.conf.in [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+use strict;
+use warnings;
+
+use OpenSSL::Test;
+use OpenSSL::Test::Utils qw(anydisabled);
+
+my $dir_sep = $^O ne "VMS" ? "/" : "";
+
+my $server = {
+    "ECDSA.Certificate" => "\${ENV::TEST_CERTS_DIR}${dir_sep}server-ecdsa-cert.pem",
```
Comment (ekasper) [inline test/ssl-tests/20-cert-select.conf.in]: This is signed by the same root cert as the RSA certificate, right?

File: test/ssl-tests/20-cert-select.conf.in [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    "ECDSA.PrivateKey" => "\${ENV::TEST_CERTS_DIR}${dir_sep}server-ecdsa-key.pem",
+    "MaxProtocol" => "TLSv1.2"
```
Comment (ekasper) [inline test/ssl-tests/20-cert-select.conf.in]: Should there be a TODO here for also testing TLS 1.3?

File: test/README.ssltest.md [hunk 3]
```diff
@@ -89,6 +89,9 @@ handshake.
 
 * ExpectedTmpKeyType - the expected algorithm or curve of server temp key
 
+* ExpectedServerKeyType, ExpectedClientKeyType - the expected algorithm or
```
Comment (ekasper) [inline test/README.ssltest.md]: Should be CertType not KeyType I think?

The client type isn't tested anywhere yet, is that right?

File: test/ssl_test.c [hunk 4]
```diff
@@ -187,17 +187,38 @@ static int check_resumption(HANDSHAKE_RESULT *result, SSL_TEST_CTX *test_ctx)
     return 1;
 }
 
-static int check_tmp_key(HANDSHAKE_RESULT *result, SSL_TEST_CTX *test_ctx)
+static int check_key_type(const char *name,int expected_key_type, int key_type)
```
Comment (ekasper) [inline test/ssl_test.c]: nit: space after `*name,`

File: test/ssl-tests/20-cert-select.conf.in [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        client => {
+		    "CipherString" => "aECDSA",
+        },
+        test   => {
+            "ExpectedServerCertType" =>, "P-256",
+            "ExpectedResult" => "Success"
+        },
+    },
+    {
+	    name => "RSA CipherString Selection",
```
Comment (ekasper) [inline test/ssl-tests/20-cert-select.conf.in]: nit: indentation looks off

File: test/ssl-tests/20-cert-select.conf.in [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    },
+    {
+        name => "RSA Signature Algorithm Selection",
+        server => $server,
+        client => {
+		    "SignatureAlgorithms" => "RSA+SHA256",
+        },
+        test   => {
+            "ExpectedServerCertType" =>, "RSA",
+            "ExpectedResult" => "Success"
```
Comment (ekasper) [inline test/ssl-tests/20-cert-select.conf.in]: Could you also test the case where we don't have a matching cert configured?

File: test/ssl-tests/20-cert-select.conf.in (hunk 1 above)
Comment (snhenson) [inline test/ssl-tests/20-cert-select.conf.in]: Yes

File: test/ssl-tests/20-cert-select.conf.in (hunk 2 above)
Comment (snhenson) [inline test/ssl-tests/20-cert-select.conf.in]: OK. That was one motivation for this: when the certificate selection is implemented for 1.3 we can test it.

File: test/README.ssltest.md (hunk 3 above)
Comment (snhenson) [inline test/README.ssltest.md]: No and we don't really support client certificate selection. I could add a minimal case to just test the client cert type is RSA in the relevant tests though.

Comment (ekasper) [inline test/README.ssltest.md]: You could also just remove ExpectClientCertType if there's no interesting use for it :)

**Pull request closed without merging**